### PR TITLE
Operational Credentials cluster: AddNOC: Set IPK.

### LINKED ISSUE #15582: Install IPK in GroupDataProvider at time of AddNOC
#### Problem
SDK impl of AddNOC does not install a real IPK to group key set ID 0 of  GroupDataProvider for current fabric, but this is required to do proper IPK by spec.

 #### Proposed Solution
AddNOC must install the default IPK by calling the GroupDataProvider.

### PR DESCRIPTION
**Pull request by rcasallas-silabs**
#### Problem
AddNOC does not install a real IPK
* Fixes [#15582](https://github.com/project-chip/connectedhomeip/issues/15582).

#### Change overview
* AddNOC modified to set the IPK

#### Testing
GroupDataProvider unit tests executed successfully.

### CODE REVIEW

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp [hunk 1]
```diff
@@ -565,6 +567,17 @@ bool emberAfOperationalCredentialsClusterAddNOCCallback(app::CommandHandler * co
     // Notify the secure session of the new fabric.
     commandObj->GetExchangeContext()->GetSessionHandle()->AsSecureSession()->NewFabric(fabricIndex);
 
+    // Set the Identity Protection Key (IPK)
+    VerifyOrExit(nullptr != groups, nocResponse = ConvertToNOCResponseStatus(CHIP_ERROR_INTERNAL));
+    VerifyOrExit(ipkValue.size() != Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES,
+                 nocResponse = ConvertToNOCResponseStatus(CHIP_ERROR_INTERNAL));
+    keyset.keyset_id     = 0;
+    keyset.policy        = chip::app::Clusters::GroupKeyManagement::GroupKeySecurityPolicy::kStandard;
+    keyset.num_keys_used = 0;
```
Comment (tcarmelveilleux) [inline src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:576]: num_keys_used = 1 ?

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp [hunk 2]
```diff
@@ -565,6 +567,17 @@ bool emberAfOperationalCredentialsClusterAddNOCCallback(app::CommandHandler * co
     // Notify the secure session of the new fabric.
     commandObj->GetExchangeContext()->GetSessionHandle()->AsSecureSession()->NewFabric(fabricIndex);
 
+    // Set the Identity Protection Key (IPK)
+    VerifyOrExit(nullptr != groups, nocResponse = ConvertToNOCResponseStatus(CHIP_ERROR_INTERNAL));
+    VerifyOrExit(ipkValue.size() != Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES,
+                 nocResponse = ConvertToNOCResponseStatus(CHIP_ERROR_INTERNAL));
```
Comment (tcarmelveilleux) [inline src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:573]: This should yield an INVALID_COMMAND status. Does CHIP_ERROR_INTERNAL map to that?

Comment (bzbarsky-apple) [reply src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:573]: It does not.  But more to the point, we should not be doing the exit to where we send a NOCResponse here at all; we should in fact be sending an error status.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp (hunk 1 above)
Comment (rcasallas-silabs) [reply src/app/clusters/operational-credentials-server/operational-credentials-server.cpp]: Fixed

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp [hunk 3]
```diff
@@ -565,6 +567,17 @@ bool emberAfOperationalCredentialsClusterAddNOCCallback(app::CommandHandler * co
...
+    keyset.keyset_id     = 0;
```
[the remark this answers is not in the document]
Comment (rcasallas-silabs) [reply src/app/clusters/operational-credentials-server/operational-credentials-server.cpp]: Comment added.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp [hunk 4]
```diff
@@ -565,6 +573,16 @@ bool emberAfOperationalCredentialsClusterAddNOCCallback(app::CommandHandler * co
     // Notify the secure session of the new fabric.
     commandObj->GetExchangeContext()->GetSessionHandle()->AsSecureSession()->NewFabric(fabricIndex);
 
+    // Set the Identity Protection Key (IPK)
+    VerifyOrExit(ipkValue.size() != Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES,
+                 nocResponse = ConvertToNOCResponseStatus(CHIP_ERROR_INVALID_ARGUMENT));
+    keyset.keyset_id     = 0; // The IPK SHALL be the operational group key under GroupKeySetID of 0
+    keyset.policy        = chip::app::Clusters::GroupKeyManagement::GroupKeySecurityPolicy::kStandard;
```
Comment (bzbarsky-apple) [inline src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:580]:
```suggestion
    keyset.policy        = GroupKeyManagement::GroupKeySecurityPolicy::kStandard;
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp (hunk 4 above)
Comment (rcasallas-silabs) [reply src/app/clusters/operational-credentials-server/operational-credentials-server.cpp]: Modified as suggested.

**Review by msandstedt — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
> When this lands, will our CASE code actually use the IPK that has been stored from AddNOC?

Based on a quick skim, no.  Need followups to do that...

**Pull request merged**